test/functional/test_framework/p2p.py

```diff
@@ -221,14 +223,15 @@ def connection_made(self, transport):
         if self.supports_v2_p2p and self.v2_state.initiating and not self.v2_state.tried_v2_handshake:
             send_handshake_bytes = self.v2_state.initiate_v2_handshake()
             self.send_raw_message(send_handshake_bytes)
-        if self.on_connection_send_msg and not self.supports_v2_p2p:
+        if self.on_connection_send_msg and not self.supports_v2_p2p and not self.reconnect:
```

Comment:
(originally because `on_connection_send_msg` would be end up being unused in v2 connection anyways and wanted to keep the `on_connection_send_msg` send during the v1 reconnection phase.)

interesting question! you're talking about a less common scenario where initiator node is preparing for v2 connection thinking responder node is v2 but v1 version message from responder node gets sent before v2 ellswift bytes from initiator node. if this happens the initiator node wouldn't attempt a reconnection using v1 since it's [receive buffer is not empty](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/net.cpp#L1525).

Reply:
I think it deserves a comment :)

Reply:
The python code only supports the "reconnection" use case in one direction: TestNode connects to the P2PConnection with v2, and reconnects with v1 without anything breaking (e.g. timeouts) on the python side. The reverse direction (P2PConnection actively does reconnections itself for outgoing connections) is not supported and I don't think it needs to be. So P2PConnection is receiving an inbound connection, and here it must wait to send `on_connection_send_msg` with sending until the version is received so it doesn't send it out in the v2 attempt. Maybe a comment could just point that in the reconnection case, `on_connection_send_msg` is sent in `on_version`, not here.

I don't understand why in the existing v1 master code, the test framework, when receiving an inbound connection, sends out the version message in `connection_made` instead of doing it in `on_version` (similar to the way it is done in bitcoind, see [here](https://github.com/bitcoin/bitcoin/blob/207220ce8b767d8efdb5abf042ecf23d846ded73/src/net_processing.cpp#L3429-L3432)). After all, the usual protocol is that the initiator sends its version, and the responder processes it and responds. But that is unrelated to this PR.

Reply:
> The reverse direction (P2PConnection actively does reconnections itself for outgoing connections) is not supported and I don't think it needs to be.

+1 since we only really care if TestNode/bitcoind (and not P2PConnection) does reconnection logic.

> I don't understand why in the existing v1 master code, the test framework, when receiving an inbound connection, sends out the version message in connection_made instead of doing it in on_version (similar to the way it is done in bitcoind, see [here](https://github.com/bitcoin/bitcoin/blob/207220ce8b767d8efdb5abf042ecf23d846ded73/src/net_processing.cpp#L3429-L3432)). After all, the usual protocol is that the initiator sends its version, and the responder processes it and responds. But that is unrelated to this PR.

i guess support for P2PConnection receiving an inbound connection/TestNode initiating outbound connections was only added later in the test framework when `addconnection` RPC was introduced. Also the concept of who's initiating wasn't relevant until now and we don't have access to that information in the inside functions.

maybe we can move [`initiating`](https://github.com/bitcoin/bitcoin/blob/fa4d9d5b39d5652de1d83564f19b1a0749863b16/test/functional/test_framework/v2_p2p.py#L67) from `class EncryptedP2PState` to `class P2PConnection` instead.